packages/jsii-pacmak/lib/generator.ts

```diff
@@ -48,18 +48,54 @@ export interface IGenerator {
   save(outdir: string, tarball: string): Promise<any>;
 }
 
+export abstract class GeneratorBase implements IGenerator {
+  abstract generate(fingerprint: boolean): void
+  abstract async load(packageDir: string, assembly: reflect.Assembly): Promise<void>
+  abstract async upToDate(outDir: string): Promise<boolean>
+  abstract async save(outdir: string, tarball: string): Promise<any>
+
+  protected calculateFingerprint(assembly: reflect.Assembly): string {
+    return crypto.createHash('sha256')
+      .update(VERSION_DESC)
+      .update('\0')
+      .update(assembly.fingerprint)
+      .digest('base64');
+  }
+
+  /**
+   * Returns the file name of the assembly resource as it is going to be saved.
+   */
+  protected getAssemblyArchiveFileName(assembly: reflect.Assembly): string {
+    const spec = assembly.spec;
+
+    let name = spec.name;
+    const parts = name.split('/');
+
+    if (parts.length === 1) {
+      name = parts[0];
+    } else if (parts.length === 2 && parts[0].startsWith('@')) {
+      name = parts[1];
+    } else {
+      throw new Error('Malformed assembly name. Expecting either <name> or @<scope>/<name>');
+    }
+
+    return `${name}@${spec.version}.jsii.tgz`;
+  }
+}
+
 /**
  * Abstract base class for jsii package generators.
  * Given a jsii module, it will invoke "events" to emit various elements.
  */
-export abstract class Generator implements IGenerator {
+export abstract class Generator extends GeneratorBase {
   private readonly excludeTypes = new Array<string>();
   protected readonly code = new CodeMaker();
   private _assembly?: spec.Assembly;
   protected _reflectAssembly?: reflect.Assembly;
   private fingerprint?: string;
 
   public constructor(private readonly options: GeneratorOptions = {}) {
+    super();
```

Comment:
Might be interesting to actually relocate the `options` property to the `GeneratorBase` class and make it `protected`.